crypto/x509v3/v3_purp.c

```diff
@@ -354,9 +354,11 @@ static void x509v3_cache_extensions(X509 *x)
     X509_EXTENSION *ex;
     int i;
 
+#ifdef tsan_ld_acq
     /* fast lock-free check, see end of the function for details. */
-    if (tsan_load((TSAN_QUALIFIER int *)&x->ex_cached))
+    if (tsan_ld_acq((TSAN_QUALIFIER int *)&x->ex_cached))
         return;
+#endif
```

Comment:
This seems to be double-checked locking with atomics (which works), but I think you want a release at the end of the function (instead of the tsan_store), and I think it should go before the unlock.

Reply:
Well, it doesn't *have to* go before the unlock, and if it doesn't, then tsan_store is sufficient, because it's right **after** CRYPTO_THREAD_unlock, which ensures that stores are committed. Of course setting ex_cached prior unlock allows lock-free test to pass earlier. However! Rationale behind leaving tsan_store at its original spot is that I figured it would look more weird prior unlock. But of course in 1.0.2 we can't have extra structure member, and so it would have to be prior unlock. And would have to be passed through separate variable, as in `temp = x->ex_flags | EXFLAG_SET; tsan_st_rel(&x->ex_flags, temp);`... So I suppose it would be appropriate to move inside...

BTW, just in case for reference, it did work on most popular platform, *cough*86, because of its rigid memory model and load being after sequence point.

Reply:
There seems to be some code that assumes certain pointers aren't set, and just overrides them. It seems not setting it while the lock is held could result in leaks.

Reply:
So I've been reading more on the doubled-checked locking, and that you just can't get it to work without the C11 memory model. See for instance https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

Reply:
And new macros are pretty much about that, i.e. turning to compiler for guarantees :-)

Reply:
The code inside the lock (after the if return) can not handle being called twice. Without the C11 memory model, the (old) if above just doesn't work, everything needs to happen inside the lock. So without atomics, the store needs to be before the unlock.

Reply:
> The code inside the lock (after the if return) can not handle being called twice.

That's why there is second check EXFLAG_SET **after** lock is acquired. T1: unlock, T2: check for ex_cached, still zero, get lock, check for EXFLAG_SET, which is set, unlock and return; T1: set ex_cache. And with store within lock it's actually more weird. T1: set ex_chached, T2: check for ex_cached, it's set, return, T1: unlock. So that T2 starts using structure before it's unlocked. It's not really a problem, but it surely looks weird. As mentioned earlier :-)

Reply:
> So that T2 starts using structure before it's unlocked [by T1]. It's not really a problem, but it surely looks weird.

Actually what is more relevant is whether or not Thread Sanitizer would consider this appropriate. I mean we can stipulate that code is correct and it's safe to use structure in question before it's unlocked, but if Thread Sanitizer complains, we would rather keep it silent (so that it won't be distraction when running into actual bugs).